File: legtool/herkulex_tool.py
#!/usr/bin/python

# Copyright 2014 Josh Pieper, jjp@pobox.com.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

'''A simple command line tool for commanding and monitoring Dongbu
HerkuleX servos.'''

import trollius as asyncio
from trollius import From, Return
import optparse
import sys
import time

from legtool.servo import herkulex

def get_address(options):
    if options.address is None:
        return herkulex.HerkuleX.BROADCAST

    return int(options.address)

@asyncio.coroutine
def do_enumerate(servo, options):
    print(yield From(servo.enumerate()))

@asyncio.coroutine
def do_set_address(servo, options):
    address = int(options.set_address)
    if address == servo.BROADCAST:
        raise RuntimeError('address cannot be set using broadcast address')
    yield From(
        servo.eep_write(get_address(options), servo.EEP_ID, [ address ]))

@asyncio.coroutine
def do_reboot(servo, options):
    yield From(servo.reboot(get_address(options)))

@asyncio.coroutine
def do_status(servo, options, do_print=True):
    addr = get_address(options)
    #status = yield From(servo.status(addr))
    position, status = yield From(servo.position(addr, return_status=True))
    result = 'Servo %d: (0x%02x, 0x%02x) %s | pos %d' % (
        addr, status.reg48, status.reg49,
        ','.join(status.active_flags()) or 'none',
        position or -1)
    if do_print:
        print result
    else:
        raise Return(result)

@asyncio.coroutine
def do_status_loop(servo, options):
    t0 = time.time()
    last_status = None
    while True:
        dt = time.time() - t0
        status = yield From(do_status(servo, options, do_print=False))
        if status != last_status:
            print '[%6.2f]' % dt, status
            last_status = status
        yield From(asyncio.sleep(0.1))

@asyncio.coroutine
def do_voltage(servo, options):
    print(yield From(servo.voltage(get_address(options))))

@asyncio.coroutine
def do_temperature(servo, options):
    print(yield From(servo.temperature_C(get_address(options))))

@asyncio.coroutine
def do_write_ram(servo, options, addr, data):
    yield From(servo.ram_write(get_address(options), addr, data))

@asyncio.coroutine
def do_blink_led(servo, options):
    addr = get_address(options)
    try:
        for _ in xrange(10):
            for leds in [servo.LED_RED, servo.LED_GREEN, servo.LED_BLUE, 0]:
                yield From(servo.set_leds(addr, leds))
                yield From(asyncio.sleep(0.2))
                sys.stderr.write(str(leds))
                sys.stderr.flush()
    finally:
        yield From(servo.set_leds(addr, 0))
        print >>sys.stderr, '0'

@asyncio.coroutine
def do_read_config(servo, options):
    result = yield From(servo.read_servo_config(get_address(options)))
    for k, v in sorted(result.items()):
        print " %-14s: %r" % (k, v)

def main():
    parser = optparse.OptionParser()
    parser.add_option('-a', '--address', default=None,
                      help='servo to communicate with')
    parser.add_option('-d', '--device', default='/dev/ttyUSB0',
                      help='serial port device')
    parser.add_option('-e', '--enumerate', action='store_true', default=None,
                      help='enumerate servos on the bus')
    parser.add_option('--set-address', dest='set_address', default=None,
                      help='set all servos on the bus to the given address')
    parser.add_option('-r', '--reboot', action='store_true', default=None,
                      help='reboot servos')
    parser.add_option('-s', '--status', action='store_true', default=None,
                      help='query status of servo')
    parser.add_option('-S', '--status-loop', action='store_true', default=None,
                      help='continuously query status of servo and print '
                      'any changes')
    parser.add_option('-v', '--voltage', action='store_true', default=None,
                      help='query voltage of servo')
    parser.add_option('-t', '--temperature', action='store_true', default=None,
                      help='query temperature of servo')
    parser.add_option('-b', '--blink_led', action='store_true', default=None,
                      help='Blink servo\'s LEDs')
    parser.add_option('-w', '--write-ram', action='append', default=[],
                      metavar='ADDR:D0:[D1..]',
                      help="Execute 'write ram' command. May be specified "
                      "multiple times")
    parser.add_option('-c', '--read-config', action='store_true', default=None,
                      help='Read and display servo config')

    (options, args) = parser.parse_args()

    if args:
        parser.error('No arguments accepted')

    actions = {
        'enumerate': do_enumerate,
        'set_address': do_set_address,
        'reboot': do_reboot,
        'status': do_status,
        'status_loop': do_status_loop,
        'voltage': do_voltage,
        'temperature': do_temperature,
        'blink_led': do_blink_led,
        'read_config': do_read_config,
        }


    write_commands = []
    for cmd in options.write_ram:
        write_commands.append(
            [int(p, 0) for p in cmd.split(':')])

    servo = herkulex.HerkuleX(options.device)

    @asyncio.coroutine
    def run_actions():
        yield From(servo.start())

        ran_any = False
        for cmd in write_commands:
            yield From(do_write_ram(servo, options, cmd[0], cmd[1:]))
            ran_any = True

        for key, callback in sorted(actions.items()):
            if getattr(options, key) is not None:
                yield From(callback(servo, options))
                ran_any = True

        if not ran_any:
            raise RuntimeError('no action specified')

    loop = asyncio.get_event_loop()
    loop.run_until_complete(run_actions())

if __name__ == '__main__':
    main()

File: legtool/src/herkulex_tool.cc
// Copyright 2015 Josh Pieper, jjp@pobox.com.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <string>

#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>
#include <boost/program_options.hpp>

#include "comm_factory.h"
#include "common.h"
#include "herkulex.h"
#include "herkulex_servo_interface.h"
#include "program_options_archive.h"

using namespace legtool;
namespace po = boost::program_options;

namespace {
struct Options {
  int address = 254;
};

typedef StreamFactory<StdioGenerator,
                      SerialPortGenerator,
                      TcpClientGenerator> Factory;
typedef HerkuleX<Factory> Servo;
typedef HerkuleXConstants HC;

struct CommandContext {
  const Options& options;
  Servo& servo;
  ServoInterface& servo_interface;
  std::string args;
  boost::asio::yield_context yield;
};
typedef std::function<void (CommandContext&)> CommandFunction;

enum CommandArg {
  kNoArgs,
  kArg,
};

struct Command {
  CommandArg args;
  CommandFunction function;
};

struct CommandText {
  std::string name;
  std::string args;
};

HC::Register ParseRegister(const std::string& args) {
  // First, try an integer register number.
  int reg_address = 0;
  int reg_size = 1;
  try {
    reg_address = boost::lexical_cast<int>(args);
    if (reg_address < 0 || reg_address > 0xfe) {
      throw std::runtime_error(
          (boost::format("Address out of range: %d") % reg_address).str());
    }
  } catch (boost::bad_lexical_cast&) {
    // OK, try a named one.
    HC constants;
    auto it = constants.ram_registers.find(args);
    if (it == constants.ram_registers.end()) {
      throw std::runtime_error(
          (boost::format("Could not parse RAM address '%s'") %
           args).str());
    }
    reg_address = it->second.position;
    reg_size = it->second.length;
  }

  return HC::Register{
    static_cast<uint8_t>(reg_address),
        static_cast<uint8_t>(reg_size)};
}

const std::map<std::string, Command> g_commands = {
  { "reboot", { kNoArgs, [](CommandContext& ctx) {
        ctx.servo.Reboot(ctx.options.address, ctx.yield);
      } } },
  { "status", { kNoArgs, [](CommandContext& ctx) {
        auto response = ctx.servo.Status(ctx.options.address, ctx.yield);
        std::cout << (boost::format("Servo %d: (0x%02x 0x%02x)\n") %
                      static_cast<int>(ctx.options.address) %
                      static_cast<int>(response.reg48) %
                      static_cast<int>(response.reg49));
      } } },
  { "voltage", { kNoArgs, [](CommandContext& ctx) {
        int value = ctx.servo.RamRead(ctx.options.address, HC().voltage(),
                                      ctx.yield);
        std::cout << (boost::format("Servo %d: %d\n") %
                      static_cast<int>(ctx.options.address) %
                      value);
      } } },
  { "ram_read", { kArg, [](CommandContext& ctx) {
        HC::Register reg = ParseRegister(ctx.args);

        auto response = ctx.servo.MemRead(
            ctx.servo.RAM_READ, ctx.options.address,
            reg.position, reg.length, ctx.yield);
        std::cout << (boost::format("Servo %d: Address %d: ") %
                      static_cast<int>(ctx.options.address) %
                      static_cast<int>(reg.position));
        for (char c: response.register_data) {
          std::cout << boost::format(" %02X") % static_cast<int>(c);
        }
        std::cout << "\n";
      } } },
  { "ram_write", { kArg, [](CommandContext& ctx) {
        std::vector<std::string> args;
        boost::split(args, ctx.args, boost::is_any_of(":"));
        if (args.size() < 2) {
          throw std::runtime_error("ram_write requires ADDR:HEXDATA");
        }

        HC::Register reg = ParseRegister(args.at(0));
        std::ostringstream ostr;
        std::string hexdata = args.at(1);
        for (size_t i = 0; i < hexdata.size(); i += 2) {
          std::string hexbyte = hexdata.substr(i, 2);
          uint8_t hexvalue = std::stoi(hexbyte, 0, 16);
          ostr.write(reinterpret_cast<const char*>(&hexvalue), 1);
        }

        ctx.servo.MemWrite(ctx.servo.RAM_WRITE, ctx.options.address,
                           reg.position, ostr.str(), ctx.yield);
      } } },
  { "set_address", { kArg, [](CommandContext& ctx) {
        int new_address = std::stoi(ctx.args, 0, 0);
        ctx.servo.EepWrite(ctx.options.address, HC::id(),
                           new_address, ctx.yield);
      } } },
  { "enumerate", { kNoArgs, [](CommandContext& ctx) {
        for (int i = 0; i < 254; i++) {
          try {
            ctx.servo.Status(i, ctx.yield);
            std::cout << boost::format("%d ") % i;
            std::cout.flush();
          } catch (boost::system::system_error& ec) {
            if (ec.code() != boost::asio::error::operation_aborted) {
              throw;
            }
          }
        }
        std::cout << "\n";
      } } },
  { "set_pose", { kArg, [](CommandContext& ctx) {
        std::vector<std::string> args;
        boost::split(args, ctx.args, boost::is_any_of(","));

        std::vector<ServoInterface::Joint> joints;
        for (std::string jstr: args) {
          std::vector<std::string> fields;
          boost::split(fields, jstr, boost::is_any_of(":"));
          joints.emplace_back(ServoInterface::Joint{
              std::stoi(fields.at(0)),
                  std::stod(fields.at(1))});
        }

        boost::asio::detail::async_result_init<
          boost::asio::yield_context,
          void (boost::system::error_code)> init(
              BOOST_ASIO_MOVE_CAST(boost::asio::yield_context)(ctx.yield));

        ctx.servo_interface.SetPose(joints, ErrorHandler(init.handler));

        init.result.get();
      } } },
};

class CommandValue : public boost::program_options::value_semantic {
 public:
  CommandValue(const std::string& name, CommandArg command_arg,
               std::vector<CommandText>* output)
      : name_(name), command_arg_(command_arg), output_(output) {}

  virtual std::string name() const { return name_; }
  virtual unsigned int min_tokens() const {
    return command_arg_ == kArg ? 1 : 0;
  }
  virtual unsigned int max_tokens() const { return min_tokens(); }
  virtual bool is_composing() const { return false; }
  virtual bool is_required() const { return false; }
  virtual bool apply_default(boost::any&) const { return false; }
  virtual void notify(const boost::any&) const {}
  virtual void parse(boost::any& value_store,
                     const std::vector<std::string>& new_tokens,
                     bool utf8) const {
    output_->push_back(
        CommandText{name_, command_arg_ == kArg ?
              new_tokens.at(0) : ""});
  }

 private:
  const std::string name_;
  const CommandArg command_arg_;
  std::vector<CommandText>* const output_;
};

int work(int argc, char** argv) {
  boost::asio::io_service service;
  Factory factory(service);
  Servo servo(service, factory);
  HerkuleXServoInterface<Servo> servo_interface(&servo);

  // Default to something moderately useful.
  servo.parameters()->stream.type = "serial";

  Options options;

  po::options_description desc("Allowable options");
  desc.add_options()
      ("help,h", "display usage mesage")
      ("address,a", po::value(&options.address)->default_value(options.address),
       "servo to communicate with")
      ;

  std::vector<CommandText> command_sequence;
  for (const auto& pair: g_commands) {
    desc.add_options()
        (pair.first.c_str(), new CommandValue(
            pair.first, pair.second.args, &command_sequence), "");
  }

  ProgramOptionsArchive(&desc, "servo.").Accept(servo.parameters());

  po::variables_map vm;
  po::store(po::parse_command_line(argc, argv, desc), vm);
  po::notify(vm);

  if (vm.count("help")) {
    std::cerr << desc;
    return 0;
  }

  boost::asio::spawn(service, ErrorWrap([&](boost::asio::yield_context yield) {
        servo.AsyncStart(yield);

        for (const auto& command: command_sequence) {
          CommandContext ctx{
            options, servo, servo_interface,
                command.args, yield};
          auto it = g_commands.find(command.name);
          BOOST_ASSERT(it != g_commands.end());
          it->second.function(ctx);
        }

        service.stop();
      }));

  service.run();
  return 0;
}
}

int main(int argc, char** argv) {
  try {
    return work(argc, argv);
  } catch (std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
  return 0;
}
